### PR TITLE
Build DoF indices in front for variables to be zeroed during residual evaluation

### LINKED ISSUE #12451: Create dof indices in front for SystemBase::zeroVariables
## Rationale

Currently the dof indices for entries to be zeroed are built on the fly in `SystemBase::zeroVariables`.
It takes up to 1/3 of cpu time in residual evaluations of one of my problem, which has lots of save-in variables.

## Description

We can build the dof indices in front probably. They can be rebuilt when mesh is changed.

## Impact
Better performance.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #12451.

### CODE REVIEW

**Comment by permcody** [comment]
This looks ok to me, @friedmud - any comments?

**Comment by permcody** [comment]
@fdkong - Can you take a look at this optimization? Make sure that you are ok with the two new closes in here. This only happens on MeshChanged so it should be a lot less frequent than doing this during the solve.

File: framework/src/systems/SystemBase.C [hunk 1]
```diff
@@ -198,7 +198,28 @@ SystemBase::zeroVariables(std::vector<std::string> & vars_to_be_zeroed)
...
+    AllLocalDofIndicesThread aldit(system(), _vars_to_be_zeroed_on_residual);
+    ConstElemRange & elem_range = *_mesh.getActiveLocalElementRange();
+    Threads::parallel_reduce(elem_range, aldit);
+
+    _dof_indices_to_zero_on_residual = aldit._all_dof_indices;
+  }
+  NumericVector<Number> & solution = this->solution();
+
+  solution.close();
+
```
Comment (fdkong) [inline framework/src/systems/SystemBase.C]: Not sure why we need a `close` here.  I would like to remote this one.

**Review by fdkong — CHANGES_REQUESTED** [review]

File: framework/src/problems/FEProblemBase.C [hunk 2]
```diff
@@ -5222,6 +5222,9 @@ FEProblemBase::meshChangedHelper(bool intermediate_change)
 
   for (const auto & mci : _notify_when_mesh_changes)
     mci->meshChanged();
+
+  _nl->clearDofIndicesToZero();
+  _aux->clearDofIndicesToZero();
```
Comment (fdkong) [inline framework/src/problems/FEProblemBase.C]: You might do the same thing for the` displaced  system` .  Create a ` _displ_sys->clearDofIndicesToZero()`

**Pull request closed without merging**